src/utils/nacl.ts

```diff
@@ -0,0 +1,94 @@
+import { randomBytes } from '@noble/ciphers/crypto';
+import { hsalsa, secretbox } from '@noble/ciphers/salsa';
+import { u32, u8 } from '@noble/ciphers/utils';
+import { x25519 } from '@noble/curves/ed25519';
+
+const PUBLIC_KEY_BYTES = 32;
+const SECRET_KEY_BYTES = 32;
+const NONCE_BYTES = 24;
+const EXPANDED_KEY_BYTES = 32;
+
+// hardcoded value of `new TextEncoder().encode('expand 32-byte k');`
+const _sigma = new Uint8Array([
+  101, 120, 112, 97, 110, 100, 32, 51, 50, 45, 98, 121, 116, 101, 32, 107,
+]);
+const _0 = new Uint8Array(16);
+
+const computeSharedKey = (pk: Uint8Array, sk: Uint8Array): Uint8Array => {
+  const s = x25519.getSharedSecret(sk, pk);
+  const k32 = new Uint32Array(EXPANDED_KEY_BYTES / 4);
+  hsalsa(u32(_sigma), u32(s), u32(_0), k32);
+  return u8(k32);
+};
```

Comment:
Interesting.. is this like a KDF?

Note - I def recommend testing this on mobile. @noble/cyphers is a JS implementation, and mobile (hermes engine) does not JIT and the performance suffers.

Mobile team has replaces SCRYPT, PBKDF2, and even AES-CBC for native libraries that get injected lol.

Reply:
Another side-note is we may need to replace AES-GCM in user-storage... but this could be replaced by the nacl box below.

Reply:
` hsalsa(u32(_sigma), u32(s), u32(_0), k32);` effectively acts as a KDF to "expand" the secret agreed upon by the ECDH step earlier. The secret is already 32 bytes so it technically shouldn't need expansion, but since it's effectively an elliptic curve point, parts of it are predictable (maybe?.. I'm not a cryptographer).

The use of hsalsa here is not my decision; this is the way `crypto_box` does it.

Reply:
Nice. Thanks for the explanation.

Reply:
🙇‍♂️ anytime